### ISSUE TITLE
more group-based splitting methods

### ISSUE THREAD

**Issue by topepo** [tidymodels/rsample#207]
It would be good to have an `initial_group_split(data, group, strata, prop)` method that can split the data when there are groups (perhaps patients). The  `strata` option might be difficult when the outcome (or other stratification variable) is not constant within each group. We could also use the median or mode on the stratification variable and use that. 

Similarly, a `mc_group_cv()` function would also be a good idea (using the splitting method as above).

**Comment by NateMietk** [comment]
I second this feature! I have pre-defined testing/training datasets so I cannot use the split_initial function, which in turn does not allow me to utilize the the last_fit function. This would open up the world of tidymodels for me and other uses like me!

**Comment by juliasilge** [comment]
@NateMietk It sounds like you may be interested in a different kind of feature, one that we already support. If you want to create an initial split from your own training/testing data, then you want to use `rsample::make_splits()`. Check out:

- [this Stack Overflow answer](https://stackoverflow.com/a/64004636/5468471)
- [this GitHub issue](https://github.com/tidymodels/rsample/issues/158)

for some examples on how to use it.

**Comment by hnagaty** [comment]
I second this too. I came across a data set that has a grouping variable. I already used `group_vfold_cv()` for the cross validation resamples, and it would be helpful to have a similar approach to the initial split.

**Comment by rkb965** [comment]
This would be a great idea. :) I'm trying to bootstrap with repeated measures data and want all measurements from an individual to be in the same split. It seems like this grouped `initial_split` would enable this, correct? Do y'all have any ideas for work arounds in the meantime? Thanks for all of the incredible work on `tidymodels`!

**Comment by juliasilge** [comment]
Yes @rkb965 in the meantime, you can create a custom split with [`make_splits()`](https://rsample.tidymodels.org/reference/make_splits.html). Check out:

- [this Stack Overflow answer](https://stackoverflow.com/a/64004636/5468471)
- [this GitHub issue](https://github.com/tidymodels/rsample/issues/158)

for some examples on how to use it.

**Comment by astamm** [comment]
Assuming that stratification variable strategies the groups, could something like have any sense:
```
group_initial_split <- function(data, group, prop = 3/4, strata = NULL, breaks = 4, pool = 0.1, ...) {
  strata <- enquo(strata)
  group <- enquo(group)
  data <- nest(data, data = !any_of(c(!!group, !!strata)))
  cell_split <- initial_split(data, prop, !!strata, breaks, pool, ...)
  train_id <- data$id[cell_split$in_id]
  cell_split$data <- unnest(cell_split$data, cols = data)
  cell_split$in_id <-  which(cell_split$data$id %in% train_id)
  cell_split
}
```

**Comment by juliasilge** [comment]
We'll want to add `group` as an argument to `mc_cv()`, `initial_split()`, and `vfold_cv()`. (Nothing else, right?)

Since `initial_split()` uses `mc_cv()` internally, let's make the changes in `mc_cv()`, specifically `mc_splits()` (along with `vfold_splits()`).

**Comment by MSHelm** [comment]
Having the `group` argument for `validation_split()` would also be very helpful for my case. Then I could directly use it for hyperparameter tuning, using e.g. `tune_grid`

**Comment by mattwarkentin** [comment]
Related: #284

**Comment by mikemahoney218** [comment]
From conversation with Julia (where we talked about this in reference to spatialsample), it seems like adding `_group_` functions would be better than adding `group` arguments. You can't specify both `group` and `strata` (plus `pool` and `breaks`), which introduces some nasty [dependencies between arguments(https://design.tidyverse.org/args-independence.html). Separate functions avoids the issue.

**Comment by mikemahoney218** [comment]
> We'll want to add `group` as an argument to `mc_cv()`, `initial_split()`, and `vfold_cv()`. (Nothing else, right?)

Looks like there's a desire for `bootstraps()` as well:
https://community.rstudio.com/t/how-to-use-rsample-for-multilevel-resampling/140337

**Comment by mattwarkentin** [comment]
@mikemahoney218 Personally, I would love to see as many of the current resampling methods as is possible gain the ability to respect the hierarchical structure while resampling. Based on a quick skim this seems like it would include: `inital_split()`, `validation_split()`, `bootstraps()`, `vfold_cv()`, `mc_cv()`, and perhaps the rolling or sliding resampling methods.

**Comment by mikemahoney218** [comment]
Howdy folks -- just to update the thread, we've implemented `group_initial_split()`, `group_validation_split()`, `group_bootstraps()`, `group_vfold_cv()` and `group_mc_cv()`. If you have a use case for a grouped variant of another resampling function, please open a new issue with a [reprex](https://reprex.tidyverse.org/) and what you'd expect that function to do!

As for stratification of grouped resamples, we've opened a new issue (#317) to try and collect some opinions on what people would _expect_ stratification with groups to do. If you've got thoughts, please let us know over there!

**Comment by Jeffrothschild** [comment]
@mikemahoney218 this is awesome and just what I need, thank you! One thing I noticed though, is that set.seed doesn't seem to be applied to the splits? 

Is it possible to have set.seed allow the group_initial_split() to separate things in the same way?

Here is an example, and if you restart R studio you'll see different colors each time.

```
library(tidyverse)
library(rsample)

df <- starwars %>% 
  mutate(name = factor(name))

set.seed(3332)
group_split <- group_initial_split(df, group = name)
group_train <- training(group_split)
group_test <- testing(group_split)

group_train %>% select(mass, name) %>% mutate(group = "train") %>% 
  bind_rows(group_test %>% select(mass, name) %>% mutate(group = "test")) %>% 
  ggplot(aes(mass, name, color = group))+
  geom_point()
```
Thanks

**Comment by mikemahoney218** [comment]
> Here is an example. It does seem to stay the same when repeating the process in the same session, but if you restart R studio you'll see different colors each time.

Opened #342 (and #343) for this. Thanks for the report!
